### ISSUE TITLE
Can't roundtrip a SDL file that contains Directives / Missing Schema declaration 

### ISSUE THREAD

**Issue by jjn1056** [graphql-perl/graphql-perl#20]
Right now if you have a SDL with directives you can parse those (with the branch associated with my PR#18) but if you serialize you don't get the directives back.  For example:

```
ok my $doc = q<
  directive @length(max: Int) on FIELD_DEFINITION | INPUT_FIELD_DEFINITION
  directive @note(msg: String) on SCHEMA | SCALAR | OBJECT | FIELD_DEFINITION 
    | ARGUMENT_DEFINITION | INTERFACE | UNION | ENUM | ENUM_VALUE
    | INPUT_OBJECT | INPUT_FIELD_DEFINITION

  type Todo {
    task: String! @length(max: 15)
  }

  type Query {
    todos: [Todo]
  }

  type Mutation @length(max: 15) {
    add_todo(task: String! @length(max: 15)): Todo
  }

  schema @note(msg: "Test") {
    query: Query
    mutation: Mutation
  }

>;

ok my $schema = GraphQL::Schema->from_doc($doc);

warn $schema->to_doc;
```

You get the following output:

```
directive @length(max: Int) on FIELD_DEFINITION | INPUT_FIELD_DEFINITION

directive @note(msg: String) on SCHEMA | SCALAR | OBJECT | FIELD_DEFINITION | ARGUMENT_DEFINITION | INTERFACE | UNION | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION

type Mutation {
  add_todo(task: String!): Todo
}

type Query {
  todos: [Todo]
}

type Todo {
  task: String!
}
```

Looks to me we just need to add serialization of directives to the right places in a bunch of ->to_doc methods.

There's a second issue, when you issue a to_doc on $schema we don't output the Schema declaration.  I notice that all the other libraries in different languages do.  I don't think it would be hard to add and I can't see the harm in it.

**Comment by mohawk2** [comment]
To deal with the second issue first: this is not in fact the case! The schema `->to_doc` is careful to only output anything if the query and mutation exist and are named other than the default (Query and Mutation). Take a look in `t/util-buildschema.t` :-)

**Comment by mohawk2** [comment]
As it is now possible to round-trip at least some directives, and your directives test has been merged in slightly beefed-up form (thanks!), I will now close this issue.

**Comment by mohawk2** [comment]
Released as 0.38! Currently the directives are invisible to the rest of the system, and they will need hooking up as actual GraphQL::Directive objects, which will involve me figuring out the API to replicate the `@skip` and `@include` directives, and enable plugins to be supplied, like with types. Suggestions welcome!
